File: dotnet/src/TunnelTanks.Core/World.cs
namespace TunnelTanks.Core;

using TunnelTanks.Core.Types;
using TunnelTanks.Core.Config;
using TunnelTanks.Core.Terrain;
using TunnelTanks.Core.Entities;
using TunnelTanks.Core.Collision;
using TunnelTanks.Core.Input;
using TunnelTanks.Core.Entities.Projectiles;
using TunnelTanks.Core.Entities.Machines;
using TunnelTanks.Core.Entities.Links;
using System.Diagnostics;

public class World
{
    private readonly Terrain.Terrain _terrain;
    private readonly TankBases _tankBases = new();
    private readonly TankList _tankList = new();
    private readonly ProjectileList _projectiles = new();
    private readonly MachineList _machines = new();
    private readonly LinkMap _linkMap = new();
    private readonly SpriteList _sprites = new();
    private readonly WorldSectors _sectors;

    private int _advanceCount;
    private TimeSpan _elapsed;
    private TimeSpan _regrowAccumulator;
    private readonly Stopwatch _regrowTimer = new();
    private bool _gameOver;

    public Terrain.Terrain Terrain => _terrain;
    public TankBases TankBases => _tankBases;
    public TankList TankList => _tankList;
    public ProjectileList Projectiles => _projectiles;
    public MachineList Machines => _machines;
    public LinkMap LinkMap => _linkMap;
    public SpriteList Sprites => _sprites;
    public WorldSectors Sectors => _sectors;
    public int AdvanceCount => _advanceCount;
    public bool IsGameOver => _gameOver;

    public SimulationProfile Profile { get; } = new();

    public World(Size terrainSize)
    {
        _terrain = new Terrain.Terrain(terrainSize);
        _sectors = new WorldSectors(terrainSize);
        _regrowTimer.Start();
    }

    public void Initialize(Terrain.Terrain generatedTerrain, Position[] spawns, int? materializeSeed = null)
    {
        for (int i = 0; i < generatedTerrain.Size.Area; i++)
            _terrain[i] = generatedTerrain[i];

        _terrain.MaterializeTerrain(materializeSeed);

        for (int i = 0; i < spawns.Length; i++)
            _tankBases.AddBase(spawns[i], i);

        _tankBases.CreateBasesInTerrain(_terrain);

        for (int i = 0; i < spawns.Length; i++)
        {
            var tankBase = _tankBases.GetSpawn(i);
            if (tankBase != null)
                _tankList.AddTank(i, tankBase);
        }
    }

    public void Advance(Func<int, ControllerOutput> getInput)
    {
        if (_gameOver) return;

        var frameWatch = Stopwatch.StartNew();

        _advanceCount++;
        _elapsed += Tweaks.World.AdvanceStep;

        { var w = Stopwatch.StartNew(); RegrowPass(); Profile.Regrow += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _projectiles.Advance(_terrain, _tankList); Profile.Projectiles += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _tankList.Advance(this, getInput); Profile.Tanks += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _machines.Advance(_terrain, Tweaks.World.AdvanceStep); Profile.Harvesters += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _sprites.Advance(Tweaks.World.AdvanceStep); Profile.Sprites += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _tankBases.Advance(); Profile.Bases += w.Elapsed; }
        { var w = Stopwatch.StartNew(); _linkMap.Advance(_terrain); Profile.Links += w.Elapsed; }

        Profile.Total += frameWatch.Elapsed;
        Profile.FrameCount++;

        if (Profile.FrameCount >= 100)
            Profile.Report();
    }

    public void SetGameOver() => _gameOver = true;

    private void RegrowPass()
    {
        if (_regrowTimer.Elapsed < _regrowAccumulator + Tweaks.World.DirtRecoverInterval)
            return;
        _regrowAccumulator += Tweaks.World.DirtRecoverInterval;

        int w = _terrain.Width, h = _terrain.Height;
        var stagedWrites = new ThreadLocal<List<(int offset, TerrainPixel value)>>(
            () => new List<(int, TerrainPixel)>(), trackAllValues: true);
        var rng = new ThreadLocal<Random>(() => new Random(), trackAllValues: true);

        Parallel.For(1, h - 1, y =>
        {
            var writes = stagedWrites.Value!;
            var random = rng.Value!;

            for (int x = 1; x < w - 1; x++)
            {
                var pos = new Position(x, y);
                var pix = _terrain.GetPixelRaw(pos);

                if (pix == TerrainPixel.Blank || Pixel.IsScorched(pix))
                {
                    int neighbors = _terrain.CountDirtNeighbors(pos);
                    int modifier = pix == TerrainPixel.Blank ? 4 : 1;
                    if (neighbors > 2 && random.Next(1000) < Tweaks.World.DirtRegrowSpeed * neighbors * modifier)
                    {
                        writes.Add((x + y * w, TerrainPixel.DirtGrow));
                    }
                }
                else if (pix == TerrainPixel.DirtGrow)
                {
                    if (random.Next(1000) < Tweaks.World.DirtRecoverSpeed)
                    {
                        var newPix = random.Next(2) == 0 ? TerrainPixel.DirtHigh : TerrainPixel.DirtLow;
                        writes.Add((x + y * w, newPix));
                    }
                }
            }
        });

        foreach (var writes in stagedWrites.Values)
        {
            foreach (var (offset, value) in writes)
            {
                _terrain.SetPixelRaw(offset, value);
                _terrain.CommitPixel(new Position(offset % w, offset / w));
            }
        }

        stagedWrites.Dispose();
        rng.Dispose();
    }
}

public class SimulationProfile
{
    public TimeSpan Regrow;
    public TimeSpan Projectiles;
    public TimeSpan Tanks;
    public TimeSpan Harvesters;
    public TimeSpan Sprites;
    public TimeSpan Bases;
    public TimeSpan Links;
    public TimeSpan Total;
    public int FrameCount;

    public void Report()
    {
        if (FrameCount == 0) return;
        Console.WriteLine($"[Profile] regrow={Avg(Regrow):F3} bases={Avg(Bases):F3} " +
            $"proj={Avg(Projectiles):F3} tanks={Avg(Tanks):F3} harv={Avg(Harvesters):F3} " +
            $"spr={Avg(Sprites):F3} links={Avg(Links):F3} " +
            $"| total={Avg(Total):F3} ms (avg over {FrameCount} frames)");
        Reset();
    }

    private double Avg(TimeSpan ts) => ts.TotalMilliseconds / FrameCount;

    public void Reset()
    {
        Regrow = Projectiles = Tanks = Harvesters = Sprites = Bases = Links = Total = TimeSpan.Zero;
        FrameCount = 0;
    }
}

File: legacy/src/world.cpp
#include "pch.h"
#include "world.h"
#include "game.h"
#include "random.h"
#include "entity.h"

namespace crust
{
World::World(Size terrain_size)
    : terrain(terrain_size), link_map(&this->terrain), projectile_list(), harvester_list(),
      tank_list(&this->terrain, &this->projectile_list), sprite_list(),
      collision_solver(&this->terrain, &this->tank_list, &this->harvester_list), sectors(SizeF(terrain_size))
{
    //this->level->OnConnectWorld(this);
}

void World::Clear()
{
    this->projectile_list.RemoveAll();
    this->tank_list.RemoveAll();
    this->harvester_list.RemoveAll();
    this->sprite_list.RemoveAll();
    this->link_map.RemoveAll();

    entity_system.Clear();
}

void World::BeginGame(Game * with_game)
{
    this->game = with_game;
    this->tank_bases.BeginGame();
    this->terrain.BeginGame();

    entity_system.Begin();
}

void World::Advance()
{
    Stopwatch<> frame_watch;

    ++this->advance_count;
    this->time_elapsed += tweak::world::AdvanceStep;

    { Stopwatch<> w; RegrowPass(); profile.regrow += w.GetElapsed(); }

    { Stopwatch<> w; this->projectile_list.Advance(&this->terrain, &this->GetTankList()); profile.projectiles += w.GetElapsed(); }
    { Stopwatch<> w; this->tank_list.for_each([=](Tank * t) { t->Advance(*this); }); profile.tanks += w.GetElapsed(); }
    { Stopwatch<> w; this->harvester_list.Advance(&this->terrain, &this->GetTankList()); profile.harvesters += w.GetElapsed(); }
    { Stopwatch<> w; this->sprite_list.Advance(&this->terrain); profile.sprites += w.GetElapsed(); }
    { Stopwatch<> w; for (TankBase & base : this->tank_bases.GetSpawns()) base.Advance(); profile.bases += w.GetElapsed(); }
    { Stopwatch<> w; this->link_map.Advance(); profile.links += w.GetElapsed(); }
    { Stopwatch<> w; this->terrain.Advance(); profile.terrain_advance += w.GetElapsed(); }
    { Stopwatch<> w; entity_system.Advance(); profile.ecs += w.GetElapsed(); }

    profile.total += frame_watch.GetElapsed();
    ++profile.frame_count;

    if (profile.frame_count >= ProfileReportInterval)
        ReportProfile();
}

void World::ReportProfile()
{
    auto avg = [&](std::chrono::microseconds acc) -> long long { return acc.count() / profile.frame_count; };
    auto ms  = [](long long us) -> long long { return us / 1000; };
    auto frac = [](long long us) -> long long { return (us % 1000); };

    long long t = avg(profile.total);
    DebugTrace<3>("[Profile] regrow=%lld.%03lld proj=%lld.%03lld tanks=%lld.%03lld harv=%lld.%03lld "
                  "spr=%lld.%03lld bases=%lld.%03lld links=%lld.%03lld terr=%lld.%03lld ecs=%lld.%03lld "
                  "| total=%lld.%03lld ms (avg over %d frames)\n",
                  ms(avg(profile.regrow)), frac(avg(profile.regrow)),
                  ms(avg(profile.projectiles)), frac(avg(profile.projectiles)),
                  ms(avg(profile.tanks)), frac(avg(profile.tanks)),
                  ms(avg(profile.harvesters)), frac(avg(profile.harvesters)),
                  ms(avg(profile.sprites)), frac(avg(profile.sprites)),
                  ms(avg(profile.bases)), frac(avg(profile.bases)),
                  ms(avg(profile.links)), frac(avg(profile.links)),
                  ms(avg(profile.terrain_advance)), frac(avg(profile.terrain_advance)),
                  ms(avg(profile.ecs)), frac(avg(profile.ecs)),
                  ms(t), frac(t),
                  profile.frame_count);
    profile.Reset();
}

void World::Draw(WorldRenderSurface & objects_surface)
{
    /* Draw everything: */
    this->projectile_list.Draw(objects_surface);
    this->tank_list.for_each([&](Tank * t) { t->Draw(objects_surface); });
    for (const TankBase & base : this->tank_bases.GetSpawns())
        base.Draw(objects_surface);
    this->harvester_list.Draw(objects_surface);
    this->sprite_list.Draw(objects_surface);
    this->link_map.Draw(objects_surface);
}

void World::SetGameOver() { this->game->GameOver(); }

void World::RegrowPass()
{
    if (!this->regrow_timer.AdvanceAndCheckElapsed())
        return;

    Stopwatch<> elapsed;

    std::vector<ThreadLocal> threadLocals;
    this->terrain.ForEachVoxelParallel(
        [this](TerrainPixel pix, SafePixelAccessor pixel, ThreadLocal * local)
        {
            if (pix == TerrainPixel::Blank || Pixel::IsScorched(pix))
            {
                int neighbors =
                    this->terrain.CountNeighborValues(pixel.GetPosition(),
                                                      [](auto voxel) { return Pixel::IsDirt(voxel) ? 1 : 0; });
                int modifier = (pix == TerrainPixel::Blank) ? 4 : 1;
                if (neighbors > 2 && local->random.Int(0, 1000) < tweak::world::DirtRegrowSpeed * neighbors * modifier)
                {
                    local->staged_writes.emplace_back(pixel.GetIndex(), static_cast<char>(TerrainPixel::DirtGrow));
                    ++local->counter_a;
                }
            }
            else if (pix == TerrainPixel::DirtGrow)
            {
                if (local->random.Int(0, 1000) < tweak::world::DirtRecoverSpeed)
                {
                    TerrainPixel new_pix = local->random.Bool(500) ? TerrainPixel::DirtHigh : TerrainPixel::DirtLow;
                    local->staged_writes.emplace_back(pixel.GetIndex(), static_cast<char>(new_pix));
                    ++local->counter_b;
                }
            }
        },
        threadLocals, WorkerCount{PhysicalCores{}});

    int holes_decayed = 0;
    int dirt_grown = 0;
    int width = this->terrain.GetSize().x;
    for (auto & tl : threadLocals)
    {
        for (auto & [offset, value] : tl.staged_writes)
        {
            this->terrain.SetVoxelRaw(offset, static_cast<TerrainPixel>(value));
            this->terrain.CommitPixel(Position{offset % width, offset / width});
        }
        holes_decayed += tl.counter_a;
        dirt_grown += tl.counter_b;
    }

    this->regrow_elapsed += elapsed.GetElapsed();
    if (this->advance_count % 100 == 1)
    {
        this->regrow_average = this->regrow_elapsed / this->advance_count;
        DebugTrace<4>("RegrowPass takes on average %lld.%03lld ms\r\n", this->regrow_average.count() / 1000,
                      this->regrow_average.count() % 1000);
    }
}

} // namespace MyNamespace